### PR TITLE
Validate API key

### PR DESCRIPTION
**Pull request by tonidero**
### Description
Android part of https://revenuecats.atlassian.net/browse/CSDK-57

With this change we log a series of situations:
- When a user introduces an API key with an old format, we log a debug message prompting them to update
- When a user introduces an API key that is not from Amazon or Google, we log an error
- When a user introduces an API key that is from Google but configures the SDK for amazon, we log an error
- When a user introduces an API key that is from Amazon but configures the SDK for google, we log an error

This behavior was tested using unit tests and manually through the `purchase-tester` app

### TODO
- [x] Check that amazon prefix is correct
- [x] Check that we want to log and not crash if invalid API key is found
- [x] Merge https://github.com/RevenueCat/purchases-android/pull/543 before this

Note that this is based on https://github.com/RevenueCat/purchases-android/pull/543

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt
Comment (aboedo) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:2097]: > It also looks like we don't have tests for this function as of yet so something to think about

Filed as https://revenuecats.atlassian.net/browse/CSDK-101, let's take care of it once this PR gets merged!

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt
Comment (aboedo) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:209]: agreed. would you mind filing a ticket for that?

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt
Comment (aboedo) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:3985]: could you also file a ticket for it?

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:113]: Right, on a first iteration I had it there. However, the `configure` method is not currently tested in Android, which would leave a gap on our test to make sure that the validation actually happened. After discussing with Andy, he suggested moving the validation to the purchases constructor, which would allow for easier testing than on the `configure` method.

Having said that, I agree it's not ideal having to add 2 extra parameters only for this and we have already created a task to add tests for the `configure` method (https://revenuecats.atlassian.net/browse/CSDK-101). So we could potentially move this back there after those tests have been added. I'm working on this task today, so we can hold this to after that's done

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt
Comment (vegaro) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:209]: Good catch, I am more of a fun of the string nomenclature. These are old tests migrated from Java so that's probably why they are still camelCase

File: strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt
Comment (vegaro) [inline strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt:27]: good touch on the `\n`s

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (vegaro) [inline common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:20]: if we move this class to the `purchases` module, we can make them internal right? maybe it's worth to do it just for that reason.

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (vegaro) [inline common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:10]: I need to double check something with you... If we don't make this class internal, is it accessible for devs? I don't think so right? They would have to do `implementation` on the common library, or we would have to be doing `api` on `common` from the `purchases` module, right? 

Sorry I am asking this but I always get confused haha

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (vegaro) [inline common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:39]: Same about making this internal and moving the class to the `common` module

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:1]: I tend to move things into `common` whenever they are needed by other modules. In this case we could have the class in the `purchases` module and have the functions as `internal` instead of `public` and visible for testing. What do you think? I don't love the modules to be fair... this is a conversation we could have haha

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:10]: No worries! And that's correct, devs shouldn't have access to this class since `purchases` uses `implementation` on `common`. Devs shouldn't have access to it unless we used `api` instead of `implementation` on the purchases module.

File: common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt:1]: Yeah, it's only needed in the `purchases` module so it's totally possible to move it there and make it `internal`. My reasoning for putting it on `common` was that I didn't want to make `purchases` too big with a lot of internal/private code and try to keep a bit of separation between the public API and internal code using modules... That said, I don't know if that's a concern and I understand that `common` naming might indicate that it's used on multiple modules so I'm ok to move it there. Let's chat today so we can decide on the structure for this kind of code.

Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/APIKeyValidator.kt]: Moved this to the `purchases` module

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: Did the refactor for the `configure` method in https://github.com/RevenueCat/purchases-android/pull/543 and made this PR point to that one. Lmk what you think!

File: strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt
Comment (tonidero) [reply strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt:27]: In all honesty, I copied the message from iOS, but yeah, it improves readability I would say 👍

File: strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt
Comment (aboedo) [inline strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt:26]: It looks like we have a double space here - one after the comma, one before the next line

Comment (tonidero) [reply strings/src/main/java/com/revenuecat/purchases/strings/ConfigureStrings.kt:26]: Good eye! @NachoSoto I copied this message from iOS, so we probably would need to fix this there as well.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt
Comment (NachoSoto) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:101]: 😍

File: purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt
Comment (NachoSoto) [inline purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt:20]: Whoa 🤯

**Review by NachoSoto — COMMENTED** [review]
Much cleaner! I love it.

File: purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt:9]: So in order to make this a compile time constant,  the options are to declare it as a top level constant (like this) or in a `companion` object inside the class. Both would work, but the companion object approach is slightly dirtier IMO.

If we had to make a public compile time constant and tie it to the class, then using a companion object would be the way to go but for 
a private one I prefer this approach. Don't know of any best practices about this though, but this answer thinks the same way I do https://stackoverflow.com/a/49977253

File: purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/APIKeyValidator.kt:20]: This annotation is really useful since it gives you warnings if you try to go against the marked visibility. Though I think we should avoid using it as much as possible, since I believe we should test classes only with their public API. In this case, since the side effects of the public API are only logging, this was much easier though.

**Review by aboedo — APPROVED** [review]
:chefskiss:

**Comment by tonidero** [comment]
Will hold merging this until https://github.com/RevenueCat/purchases-android/pull/543 is reviewed as well, since this is based on that PR @aboedo

**Pull request merged**
